경북대 BE 이원준 3단계 - 상품 옵션

lwj0831 opened the pull request:
## ✅ 구현 사항

### 2단계 리팩토링

1. 기존 클라이언트 개발자를 위한 예외 디버깅 정보를 위한 처리 일관성 있게 변경
2. `@PageableDefault` 대신 `PageRequestDto`를 사용하여 정렬 필드 유효성까지 포함한 검증을 서비스 계층에서 수행하도록 변경 (현재 진행 중)

### 3단계 - 상품 옵션

1. WIshItem 삭제 시 리소스 소유권 검증 로직 추가
2. ProductOption 요구 사항 구현
   - Controller, Service, Repository 계층 구성
   - 도메인 제약사항 기반 예외 처리
   - `ProductOptionValidationService`를 통한 검증 책임 분리
   - 단위 및 통합 테스트 작성
3. Product 등록 시에도 최소 1개의 옵션이 포함되도록 검증 및 관련 로직 리팩토링

---
## ✅ 이 부분을 중점적으로 리뷰해주세요 🙏

1. **페이징 정렬 필드에 대한 검증 리팩토링**
- https://github.com/next-step/spring-gift-enhancement/pull/260#discussion_r2218540440

    멘토님의 전단계 피드백에 따르면 자체 DTO를 사용하여 MethodArgumentResolver와 같은 곳에서 역직렬화 시 정렬 필드에 대한 검증을 수행하는 방식을 제안주셨는데  
페이징 정렬 필드는 도메인별로 다를 수 있어서 (예를 들어, product는 id, name, price / wishItem은 id, name) 공통으로 사용하는 PageReqeustDto의 MethodArgumentResolver에서 정렬 필드명 유효성 검증을 하기엔 어렵다고 판단이 되고 도메인마다 별도의 DTO와 Resolver를 만들기에는 비효율적인 구조라 생각하여 고민중입니다. 🤔
구체적인 리팩토링 방식에 대한 조언 부탁드립니다!

2. **옵션 수량 변경 API 설계 방향**

   옵션 수량 변경에 대한 로직이 내부적으로는 서비스 단에서  `addProductOption()`, `subtractProductOption()`으로 명확히 분리한 상황입니다.
   외부 API 관점에서 옵션 수량 증가/감소를 단일 API로 통합할지, 별도 API로 나눌지에 대한 기준을 명확히 하고 싶습니다.  
   멘토님의 설계 기준이나 경험에 기반한 의견을 듣고 싶습니다.

3. **검증 로직의 위치 및 구조**

   서비스 레이어에 검증 로직이 누적되며 가독성 저하가 발생해, `ProductOptionValidationService`로 분리해 관리하고 있습니다.  
   이와 같은 구조가 괜찮은지, 또는 더 좋은 책임 분리 방식이나 구조적 대안이 있다면 공유 부탁드립니다.

lwj0831 committed:
chore: default프로파일 dev프로파일 분리

- default(H2) : 리뷰어 애플리케이션 실행 환경
- dev(MYSQL) : 개발 환경
- test(H2) : 테스트 환경 - 스프링 시큐리티 인증 필터 무효화

lwj0831 committed:
refactor: 예외 처리 방식 리팩토링

- BusinessException 내 예외 추가 정보 담는 필드 추가
- 관련 사항 변경

lwj0831 committed:
chore(wishlist): 위시 아이템 삭제시 리소스 소유권 검증 로직 추가

lwj0831 committed:
feat(product): ProductOption관련 예외 정의 및 핸들러 구현

lwj0831 committed:
feat(product): ProductOption 도메인 정의 및 JPA레포지토리 추가

lwj0831 committed:
feat(product): ProductOptionService 추가

ProductOption관련 검증 코드 ProductOptionValidationService로 분리

lwj0831 committed:
feat(productOption): ProductOptionController 추가

lwj0831 committed:
fix(admin): ProductAdminController 수정

lwj0831 committed:
refactor(product): 상품 등록 시 최소 1개 이상의 옵션 등록하도록 리팩토링

lwj0831 committed:
chore(product): ProductSortField 불필요한 메서드 제거

lwj0831 committed:
chore(test): 변경사항에 따른 ProductApiStructuredE2ETest 수정

lwj0831 committed:
feat: ProductOption 컨트롤러, 서비스 단위 테스트 추가

lwj0831 committed:
fix(productOption): 패턴 정규식 한글자만 허용하는 문제 해결

lwj0831 committed:
fix(productOption): 로직 오류 수정

lwj0831 committed:
chore: product_option 더미데이터 추가

lwj0831 committed:
docs: STEP3 3단계 README 작성

lwj0831 committed:
chore(ProductOption): ProductOption에 (product_id,name) 유니크 제약조건 추가

lwj0831 committed:
fix: SQL 문법 에러 수정 및 ProductOption 필드 nullable=false 추가

lwj0831 committed:
feat(productOption): ProductOption 개별 등록에서 Bulk Insert로 변경

lwj0831 committed:
feat(productOption): ProductOption 생성 시 name 요청값 중복 검증 추가

- 커스텀 어노테이션과 AOP기반 빈 Validation을 활용하여 구현
- DTO에서 입력값 `유효성 검증의 책임 부여

lwj0831 committed:
chore: 검증 예외 발생 시 API응답 내용 수정

lwj0831 committed:
test(productOption): ProductOptionControllerTest에 변경사항 반영

Rok93 commented on src/main/java/gift/product/controller/ProductOptionController.java:
```diff
@@ -0,0 +1,49 @@
+package gift.product.controller;
+
+import gift.global.common.dto.PageResponseDto;
```
아래 질문에 대한 답변은 Product의 페이지네이션 관련 로직들이 이번 변경에는 없어서, 적당히 코멘트 남길 곳이 없네요. 😅
여기에 남겨보겠습니다. 

> 멘토님의 전단계 피드백에 따르면 자체 DTO를 사용하여 MethodArgumentResolver와 같은 곳에서 역직렬화 시 정렬 필드에 대한 검증을 수행하는 방식을 제안주셨는데
페이징 정렬 필드는 도메인별로 다를 수 있어서 (예를 들어, product는 id, name, price / wishItem은 id, name) 공통으로 사용하는 PageReqeustDto의 MethodArgumentResolver에서 정렬 필드명 유효성 검증을 하기엔 어렵다고 판단이 되고 도메인마다 별도의 DTO와 Resolver를 만들기에는 비효율적인 구조라 생각하여 고민중입니다. 🤔
구체적인 리팩토링 방식에 대한 조언 부탁드립니다!

제가 제안드린 방식은 자체 DTO를 `MethodArgumentResolver`에 등록하여 검증하는 방식이 아니라 
DTO는 단순히 요청 객체로 말씀을드렸던거고 이 요청 객체 안의 필드로 ProductSortField, Direction와 같은 `enum class`를 파라미터 타입으로 둬서 자동으로 enum class로 역직렬화되도록 만들자는 이야기였어요. 

MethodArgumentResolver와는 상관이 없었는데, 조금 내용 전달을 잘못드렸던 것 같아요. 🙏

이런 의도의 답변이었을 때는 어떻게 생각하시는지 다시 한번 생각 공유 부탁드릴게요. 😃

lwj0831 replied:
저도 매직스트링을 사용하는 방식보다 안정성이 높은 Enum을 정의하는 방식이 적합하다고 생각합니다!

@Rok93 
도메인마다 다른 정렬 필드를 가지되, 확장성과 일관성을 갖는 구조를 고민중입니다. 저는 전략패턴으로 해결하는 방식을 생각중인데 멘토님의 의견도 한번 여쭤보고 싶습니다! 🧐

Rok93 replied:
음... 조금 조심스럽네요. 
이 또한 섣부른 추상화가 될 수 있다고 생각해요. 
생각보다 페이지네이션이 필요한 경우가 많지않을 수도 있어요. 🤔

또한 정렬이라는 것이 생각보다 특정 필드로만 오름차순(or 내림차순)한다면 정~~~말 심플하겠지만, 
복합 정렬 조건을 요구하는 경우나 혹은 계산된 값을 기준으로 정렬한다거나... 등의 요구사항이 생길 수도 있을거에요. 😅

아마 생각하시는 것보다는 단순하게 추상화가 안될 것 같다는 생각이드는 것 같아요. 😃 
저는 정렬 케이스는 중복처럼 보이더라도 분리하는 방식도 괜찮을 것 같다고 생각되네요. 🤔

Rok93 commented on src/main/java/gift/product/service/ProductService.java:
```diff
@@ -59,6 +60,13 @@ public Long createProduct(CreateProductRequestDto dto) {
             dto.imageUrl()
         );
 
+        dto.options()
+            .forEach((optionDto) -> {
+                ProductOption option = ProductOption.of(optionDto.name(), optionDto.quantity(),
+                    newProduct);
+                newProduct.addOption(option);
+            });
+
```
객체를 생성한 뒤에 객체에 옵션을 추가해주기보다는, 객체를 생성하는 시점에 Option을 같이 생성자 파라미터로 전달하여 초기화해주면 어떨까요? 

그렇다면 [미션 소개 페이지](https://edu.nextstep.camp/s/UH6H8Iba/ls/dLmzLLkv)에서 요구하고있는 아래의 요구사항도 생성 시점에 바로 검증할 수 있지 않을까요? 

> 상품에는 항상 하나 이상의 옵션이 있어야 한다

가능하면 이런 도메인 정책들을 도메인 객체에 녹이는 것이 좋은 코드라고 생각해요. 😃

lwj0831 replied:
연관관계의 주인 클래스(ProductOption)의 연관 객체 필드(Product)를 설정하는 데에만 초점을 둬서 이 부분에 대해서는 고려를 못했습니다..!
생성자에서 List<ProductOption>을 추가하고 연관관계 편의메서드를 통해 주인 클래스의 연관 객체 필드를 setting을 하고 cascadeType.PERSIST로 Product persist시 연관된 option도 persiste되도록 하였고 검증로직까지 추가하는 방식으로 리팩토링하였습니다!

@Rok93 
상품에는 항상 하나 이상의 옵션이 있어야 한다를 저는 
```
public record CreateProductOptionListRequestDto(
    @Valid
    @NotEmpty(message = "상품은 하나 이상의 옵션을 가져야 합니다.")
    @UniqueOptionNames
    List<CreateProductOptionRequestDto> optionRequestDtoList
) {

}
```
로 DTO에서 검증하고 있었는데 그렇다면 이 검증 어노테이션은 이제 제거하는 게 좋을지 아니면 그대로 두어 두 곳에서 검증을 수행하는 게 좋을지 고민입니다!
제 생각은 동일한 검증 로직이 중복이라 불필요하다고 느껴지기도 하고 해당 검증 로직에 대한 유지보수 관점에서도 동일한 검증이 애플리케이션 코드 두 곳에 있어 관리하기도 힘들 것이라 판단하여 제거하는 쪽이 좋다고 생각했습니다..!

Rok93 replied:
말씀하신 것처럼 이런 도메인 정책들을 자꾸 DTO 단에서도 검증하게되면 추후 도메인 정책이 변경되었을 때, 유지보수를 이중으로 해줘야하는 불편함이 생기게되죠. 😅  동일한 정책이 여기저기 퍼져있게되면 추후 실수할 여지도 늘어나게되니 복잡한 정책일수록 도메인 로직에만 두는 것이 좋다고 생각해요. 🤗

좋은 고민과 좋은 결과로 이어졌네요. 👍💯

Rok93 commented on src/main/java/gift/product/domain/Product.java:
```diff
@@ -22,9 +26,17 @@ public class Product {
     private String description;
     private String imageUrl;
 
+    @OneToMany(cascade = CascadeType.PERSIST, orphanRemoval = true)
```
cascade, orphanRemoval 조건도 추가해주셨군요? 💯
각각의 설정이 어떤 역할을 하는지도 학습하셨을 것 같은데, 각각의 설정은 어떤 역할을 하나요? 😃

lwj0831 replied:
cascade는 부모->자식 또는 자식->부모 엔티티로 동일한 영속상태를 적용하는 역할을 하고
orphanRemoval는 부모-자식 엔티티간의 연관관계 참조가 끊길 시(JoinColumn으로 연관관계의 주인 필드 역할을 하는 곳의 참조가 null이 되거나 다른 엔티티로 변경되는 등) DB에서 자식 엔티티가 삭제되는 걸로 알고 있습니다!

CascadeType.PERSIST를 통해 Product가 persist시 ProductOption도 자동으로 persist되어 INSERT 쿼리가 지연쿼리저장소에 담기도록 하였고
orpahanRemoval = true를 통해 연관관계가 끊기면 자식 객체인 ProductOption을 트랜잭션 커밋 시점에 스냅샷과 비교하여 orphan여부를 판단하여 delete()를 실행하여 DELETE 쿼리가 수행되도록 의도했습니다

Rok93 commented on src/main/java/gift/product/domain/ProductOption.java:
```diff
@@ -0,0 +1,96 @@
+    name = "product_option",
+    uniqueConstraints = @UniqueConstraint(
+        name = "uk_product_option_product_id_name",
+        columnNames = {"product_id", "name"}
```
복합 Unique Key 활용 💯👍

Rok93 commented on src/main/java/gift/product/domain/ProductOption.java:
```diff
@@ -0,0 +1,96 @@
+        if (result >= QUANTITY_MAX_VALUE) {
+            this.quantity = result;
+        } else {
+            throw new IllegalArgumentException("상품 수량 최대값(100000000) 초과입니다.");
```
최대값이 추후 수정된다면 위의 상수와 이 에러 메시지도 같이 수정해야할 것 같은데요. 
아래와 같이 변경한다면 추후 최댓값이 변경되면 상수만 변경하면 되지 않을까요? 🤗 

```suggestion
            throw new IllegalArgumentException("상품 수량 최대값(" +  QUANTITY_MAX_VALUE + ") 초과입니다.");
```

Rok93 commented on src/main/java/gift/product/domain/ProductOption.java:
```diff
@@ -0,0 +1,96 @@
+        return new ProductOption(null, name, quantity, product);
+    }
+
+    public static ProductOption withId(Long id, String name, int quantity, Product product) {
```
해당 정적 팩토리 메서드는 사용하지 않는 것 같은데요. 🤔
사용하지 않는다면 제거해주세요. 🙏

Rok93 commented on src/main/java/gift/product/validation/UniqueProductOptionNameValidator.java:
```diff
@@ -0,0 +1,23 @@
+import java.util.List;
+import java.util.Set;
+
+public class UniqueProductOptionNameValidator implements
+    ConstraintValidator<UniqueOptionNames, List<CreateProductOptionRequestDto>> {
```
요청부터 꼼꼼하게 잘 검증해주셨지만, 개인적인 생각으로는 요청단에서는 간단한 유효성 검증 정도만 진행하고 이런 도메인 정책들은 도메인 로직으로 유효성을 검증해주는게 좋다고 생각해요. 😃 (지금처럼 이런 검증을 모두 Controller 단에서 해주게된다면 매번 이런 Validator를 생성해줘야하기때문에 번거롭고 또 이를 유지보수하는 것도 어려워진다고 생각해요)

Option을 추가하거나 혹은 수정 혹은 새로운 상품 생성(w. 옵션)을 하는 경우 중복되는 옵션이 없는지 유효성 검증을 진행해준다면

lwj0831 replied:
도메인 내에서의 검증으로 구현할 생각을 못하고 있었습니다..! 기본적인 부분인데 너무 복잡하게 구현한 것 같아요😂
해당 방식으로 리팩토링하였습니다!

Rok93 commented on src/main/java/gift/product/service/ProductOptionValidationService.java:
```diff
@@ -0,0 +1,23 @@
+import org.springframework.stereotype.Service;
+
+@Service
+public class ProductOptionValidationService {
```
> 검증 로직의 위치 및 구조
서비스 레이어에 검증 로직이 누적되며 가독성 저하가 발생해, ProductOptionValidationService로 분리해 관리하고 있습니다.
이와 같은 구조가 괜찮은지, 또는 더 좋은 책임 분리 방식이나 구조적 대안이 있다면 공유 부탁드립니다.

물론 검증 로직이 너무 많다면 지금처럼 별도의 검증 역할을하는 객체를 뽑아보는 방법도 충분히 고려해볼 수 있을 것 같아요. 
다만 지금처럼 `xxxService`와 같은 네이밍을 쓰기보다는  `xxxValidator` 와 같이 명시적으로 Service와는 다른 역할을하는 객체라는 것을 네이밍적으로 드러낼 것 같아요. (아예 패키지를 나누는 것도 괜찮을 것 같네요) 

다만 지금과 같은 검증 로직은, 도메인 객체에게 위임해볼 수 있을 것 같은데요. 
예를들어 product 객체에게 `옵션을 더하라는 메시지`(= `addOption`)를 보냈을 때, 내부적으로 추가할 옵션과 동일한 이름을 가진 옵션이 있는지 여부를 확인해서 유효성 검증을 하면 되지 않을까요? 

도메인 객체가 충분히 직접 처리할 수 있는 책임을 Service 단에 위임하신게 아닌가 싶어요. 😃
가능하면 이런 비즈니스 로직을 Product, Option과 같은 도메인 객체에게 위임해보면 좋을 것 같습니다.

lwj0831 replied:
@Rok93 
해당 검증 로직은 DB에 저장된 ProductOption엔티티 중 이름이 중복되는지 확인하는 검증이라 도메인 객체내에 두면 도메인 클래스가 Repository를 의존하게 될 것 같아 Validator로 네이밍만 변경하였습니다

멘토님께서 말씀하신 것은 Product클래스의  ```List<ProductOption>```에 새로 추가할 옵션들과 현재 List 필드에 저장된 객체들의 이름 중복 여부를 체크하는 검증을 말씀하신 건가요??

Rok93 replied:
네 product가 상태로 `options`를 가지고 있기 때문에 상품의 기준에서 이런 유효성 검증이 가능하지 않을까요? 
상품 `객체에게 옵션을 추가해` 혹은 `특정 옵션을 변경해`와 같은 메시지를 보냈을 때, 내부적으로 중복 옵션에 대한 유효성 검증이 가능하지 않을까요?

Rok93 commented on src/main/java/gift/product/dto/CreateProductOptionRequestDto.java:
```diff
@@ -0,0 +1,24 @@
+    @Size(max = 50, message = "옵션 이름은 최대 50자까지 입력할 수 있습니다.")
+    @Pattern(
+        regexp = "^[\\p{L}\\p{N} ()\\[\\]\\+\\-\\&/_]*$",
+        message = "허용되지 않은 특수 문자가 포함되어 있습니다."
```
옵션의 이름 조건이나 수량 조건도 모두 도메인 정책이락 볼 수 있지 않을까요? 😃
ProductOption 객체가 생성되는 시점에 이런 도메인 로직들을 검증해준다면 더 풍부한 도메인 객체가 될 것 같네요. 😉

lwj0831 replied:
네 저도 해당 부분은 도메인 정책과 관련된 부분이라 DTO에서의 어노테이션 검증이 아닌 도메인 로직으로 분리하는 것이 좋다고 생각합니다!
관련해서 리팩토링하였습니다!

Rok93 commented on src/main/java/gift/product/service/ProductOptionService.java:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    @Transactional
+    public void addProductOptionQuantity(Long optionId, int quantity) {
+        ProductOption option = findProductOptionOrThrow(optionId);
+
+        option.addQuantity(quantity);
+    }
+
+    @Transactional
+    public void subtractProductOptionQuantity(Long optionId, int quantity) {
+        ProductOption option = findProductOptionOrThrow(optionId);
+
+        option.subtractQuantity(quantity);
+    }
```
> 옵션 수량 변경에 대한 로직이 내부적으로는 서비스 단에서 addProductOption(), subtractProductOption()으로 명확히 분리한 상황입니다.
외부 API 관점에서 옵션 수량 증가/감소를 단일 API로 통합할지, 별도 API로 나눌지에 대한 기준을 명확히 하고 싶습니다.
멘토님의 설계 기준이나 경험에 기반한 의견을 듣고 싶습니다.

일단 3단계 미션 요구사항에는 `상품 옵션의 수량을 지정된 숫자만큼 빼는 기능을 구현한다`의 내용만 있었던 것 같아서, 옵션 수량을 더하는 것은 요구사항이 아니였지 않았을까 싶어요. 👀

저는 더하나 빼나 비슷한 기능으로 여겨져서, `상품 수량을 수정하는 기능`으로 퉁칠 것 같기도 하네요. 🤔
당장에는 꼭 쪼개야할 이유를 잘 모르겠달까요. 😅 

[소프트웨어 3대 원칙](https://blog.naver.com/finway/221163007357) 중 `YAGNI`라는 표현을 개인적으로 좋아하는데요. 
너무 먼 미래를 생각하며 구현하는 것이 오히려 섣부른 추상화가되는 경우도 많았던 것 같아요. 
그래서 추후 수정이 그렇게 어렵지 않은 기능이라면 저는 선택을 뒤로 미룰 것 같아요. 
사용하다보면 더 명확한 요구사항이 나오는 경우도 많기 때문에 당장에 필요한 것을 이 기능을 사용하면서 얻은 피드백을 바탕으로 필요하다면 분리하고 아니라면 한 개의 API로 계속 유지할 것 같아요.

lwj0831 replied:
평소에 확장성을 지나치게 고려해서 설계해 현재 요구사항에는 적합하지 못한 선택을 많이 했었던 것 같은데 앞으로는 필요성이 느껴질 때 적용하는 방식으로 개발해보도록 하겠습니다 😊
좋은 말씀 해주셔서 감사합니다!

Rok93 reviewed (changes requested):
원준님 3단계 구현 잘해주셨네요. 👍👍

몇몇 코멘트도 남겨두었으니 같이 확인해서 반영 부탁드릴게요. 🙏 
추가적으로 질문주신 내용에대한 답변과 추가 질문거리 남겨두었는데 같이 확인 부탁드릴게요. 🤗 
다음 리뷰 요청 때는 상품 고도화 미션은 이만 종료하도록하겠습니다. 😉

lwj0831 committed:
refactor: Product 생성자에서 옵션 목록 추가 및 중복 검증 로직 도입

lwj0831 committed:
chore: ProductOptionValidator 리네이밍

lwj0831 committed:
refactor: addOption시 상품옵션 이름 검증 추가

lwj0831 committed:
chore: dto 리네이밍

lwj0831 committed:
refactor(productOption): 상품 옵션명 유효성 검증 도메인 로직 정의

lwj0831 committed:
refactor: ProductOption 검증 로직 수정

lwj0831 committed:
chore: IllegalException발생 시 커스텀 에외메세지 반환할 수 있도록 헬퍼메서드 추가

lwj0831 committed:
test: ProudctOption 도메인 검증 테스트 추가 및 변경사항에 맞게 기존 테스트 수정

lwj0831 committed:
refactor: 상품 옵션 리스트 등록 시 요청 옵션명 중복 검증 로직 추가

- ProductOptionValidator메서드로 검증 헬퍼 메서드 추가
- 요청 DTO간 검증, DB 엔티티간 검증 분리를 위해
  ProductOptionNameAlreadExistsException(DB),
  DuplicateProductOptionNameException(요청DTO)로 구분
- DB 엔티티와 중복 옵션명 검증 시 bulk연산으로 쿼리 N+1->1로 개선

lwj0831 committed:
chore: ProductOptionService 테스트 코드 수정

Rok93 reviewed (approved):
원준님 마지막까지 피드백 반영 잘해주셨습니다. 👍💯
마지막 단계도 훌륭하게 잘 진행해주셔서 상품 고도화 미션은 이만 종료하도록하겠습니다. 🎉🎉 
지금까지 잘 해오셨으니 남은 미션도 충분히 잘 마무리지으실거라 생각됩니다. 🤗
마지막 미션도 화이팅입니다. 💪💪💪
